File: idealingua/idealingua-runtime-rpc-typescript/src/main/resources/runtime/typescript/irt/either.ts

export type Either<L, A> = Left<L, A> | Right<L, A>;

export class Left<L, A> {
    constructor(readonly value: L) {
    }

    public map<B>(f: (a: A) => B): Either<L, B> {
        return this as any;
    }

    public bimap<V, B>(g: (a: A) => B, f: (l: L) => V): Either<V, B> {
        return new Left(f(this.value));
    }

    public fold<B>(whenRight: (a: A) => B, whenLeft: (l: L) => B): B {
        return whenLeft(this.value);
    }

    public bifold<V, B>(whenRight: (a: A) => B, whenLeft: (l: L) => V): V | B {
        return whenLeft(this.value);
    }

    public getOrElse(a: A): A {
        return a;
    }

    public isLeft(): this is Left<L, A> {
        return true;
    }

    public isRight(): this is Right<L, A> {
        return false;
    }

    public swap(): Either<A, L> {
        return new Right(this.value);
    }

    public filterOrElse(p: (a: A) => boolean, zero: L): Either<L, A> {
        return this;
    }

    public match(whenRight: (value: A) => void, whenLeft:  (value: L) => void): void {
        whenLeft(this.value);
    }
}

export class Right<L, A> {
    constructor(readonly value: A) {
    }

    public map<B>(f: (a: A) => B): Either<L, B> {
        return new Right(
            f(this.value)
        );
    }

    public bimap<V, B>(g: (a: A) => B, f: (l: L) => V): Either<V, B> {
        return new Right<V, B>(
            g(this.value)
        );
    }

    public fold<B>( whenRight: (a: A) => B, whenLeft: (l: L) => B): B {
        return whenRight(this.value);
    }

    public bifold<V, B>(whenRight: (a: A) => B, whenLeft: (l: L) => V): V | B {
        return whenRight(this.value);
    }

    public getOrElse(a: A): A {
        return this.value;
    }

    public isLeft(): this is Left<L, A> {
        return false;
    }

    public isRight(): this is Right<L, A> {
        return true;
    }

    public swap(): Either<A, L> {
        return new Left(this.value);
    }

    public filterOrElse(p: (a: A) => boolean, zero: L): Either<L, A> {
        return p(this.value) ? this : left(zero);
    }

    public match(whenRight: (value: A) => void, whenLeft:  (value: L) => void): void {
        whenRight(this.value);
    }
}

const of = <L, A>(a: A): Either<L, A> => {
    return new Right<L, A>(a);
};

export const left = <L, A>(l: L): Either<L, A> => {
    return new Left(l);
};

export const right = of;

File: idealingua/idealingua-runtime-rpc-c-sharp/src/main/resources/runtime/csharp/IRT/Either.cs

using System;

namespace IRT {
    public abstract class Either<LT, RT> {
        private Either() {}
        public abstract Either<LT, B> Map<B>(Func<RT, B> f);
        public abstract Either<V, B> BiMap<V, B>(Func<RT, B> g, Func<LT, V> f);
        public abstract B Fold<B>(Func<RT, B> whenRight, Func<LT, B> whenLeft);
        public abstract RT GetOrElse(RT a);
        public abstract bool IsLeft();
        public abstract bool IsRight();
        public abstract Either<RT, LT> Swap();
        public abstract Either<LT, RT> FilterOrElse(Func<RT, bool> p, LT zero);
        public abstract void Match(Action<RT> whenRight, Action<LT> whenLeft);

        public sealed class Left<L, A>: Either<L, A> {
            private readonly L Value;
            public Left(L value) {
                Value = value;
            }

            public override Either<L, B> Map<B>(Func<A, B> f) {
                return new Left<L, B>(Value);
            }

            public override Either<V, B> BiMap<V, B>(Func<A, B> g, Func<L, V> f) {
                return new Left<V, B>(f(Value));
            }

            public override B Fold<B>(Func<A, B> whenRight, Func<L, B> whenLeft) {
                return whenLeft(Value);
            }

            public override A GetOrElse(A a) {
                return a;
            }

            public override bool IsLeft() {
                return true;
            }

            public override bool IsRight() {
                return false;
            }

            public override Either<A, L> Swap() {
                return new Right<A, L>(Value);
            }

            public override Either<L, A> FilterOrElse(Func<A, bool> p, L zero) {
                return this;
            }

            public override void Match(Action<A> whenRight, Action<L> whenLeft) {
                whenLeft(Value);
            }
        }

        public sealed class Right<L, A>: Either<L, A> {
            private readonly A Value;
            public Right(A value) {
                Value = value;
            }

            public override Either<L, B> Map<B>(Func<A, B> f) {
                return new Right<L, B>(f(Value));
            }

            public override Either<V, B> BiMap<V, B>(Func<A, B> g, Func<L, V> f) {
                return new Right<V, B>(g(Value));
            }

            public override B Fold<B>(Func<A, B> whenRight, Func<L, B> whenLeft) {
                return whenRight(Value);
            }

            public override A GetOrElse(A a) {
                return Value;
            }

            public override bool IsLeft() {
                return false;
            }

            public override bool IsRight() {
                return true;
            }

            public override Either<A, L> Swap() {
                return new Left<A, L>(Value);
            }

            public override Either<L, A> FilterOrElse(Func<A, bool> p, L zero) {
                if (p(Value)) {
                    return this;
                }

                return new Left<L, A>(zero);
            }

            public override void Match(Action<A> whenRight, Action<L> whenLeft) {
                whenRight(Value);
            }
        }
    }
}
